utils/swcert-create.1
.\" swcert-create - creates a sip tls certificate pair
.\" Copyright (C) 2017-2018 Tycho Softworks
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc.,59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.\" This manual page is written especially for Debian GNU/Linux.
.\"
.TH swcert-create "1" "December 2017" "SipWitchQt" "Tycho Softworks"
.SH NAME
swcert-create \- generates a sipwitch tls certificate pair
.SH SYNOPSIS
.B swcert-create
.RI [ options ]
.br
.SH DESCRIPTION
This is used to create a self-signed public/private certificate pair for SipWitchQt 
sip tls support.  
.SH OPTIONS
.TP
.B \-\-server
Server hostname this certificate and the tls session will be associated with.
.TP
.B \-\-country
Two character country abbreviation this server is operating in.
.TP
.B \-\-organization
The organizing issuing this certificate.
.TP
.B \-\-userdata
Uses the local userdata testing path for the local sqlite database.
.SH AUTHOR
.B swcert-create
was written by David Sugar <tychosoft@gmail.com>.
.SH "REPORTING BUGS"
Report bugs to tychosoft@gmail.com.
.SH COPYRIGHT
Copyright \(co 2017 Tycho Softworks.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.



CHANGELOG
v0.2.7
* Improved packaging support
* Updated copyright notices
* Utils use stand-alone to init mysql servers
* De-emphasizing sqlite beyond testing builds

v0.2.6
* Simplified expiration handling
* Re-arranged font selection
* Introduced desktop & audio notify setting
* Integrated login auto start behavior
* Speed dial ui support
* Call coverage setting support
* Call forwarding setting support
* Delay ringing option supported
* Per user and global auto answer settings
* Start of integrated voice mail
* Device keys introduced

v0.2.5
* Customization at build time
* Topic management in sessions
* Policy setting for main public room

v0.2.4
* Session/state based toolbar management
* Online status of all extensions shown
* Membership management for groups
* Server system notification messaging
* Delete authorizations supported

v0.2.3
* Schema changes
* Message expiration

v0.2.2
* Change password support
* Simple markdown support
* Zeroconfg support
* Message sync for existing extensions
* Functional testing support
* Improved adduser ui
* Expiration marking on client

v0.2.1
* Fix for crash sending offline messages
* Visual styling improvements and optimizations

V0.2.0
* IPL loading of tables, such as for external mysql db
* User classification of priviledges
* Ability to edit extension names from profile
* Clickable and highlighted urls in messages
* Offline message sync
* Cleaner db handling for login/logout and multiple dbs
* Client import/export
* Font picker for message view
* Admin can create users from client
* Phonebook expanded
* New message indicators

v0.1.4
* I want to send a message to someone
* I Want to receive messages
* I want to search my messages
* I want to logout, and back in
* I want to know what extensions I can login as
* I want the ui to become available in language of my choice
* I want to know what devices are on my extension
* I want to autostart when I login on linux
* sipwitchqt message laying and rendering engine introduced
* email and public uri properties added to contacts

v0.1.3
* Roster processing on client and server
* Search processing on sessions screen
* click feedback, doubleclick contacts
* vibrant and light icon schemes

v0.1.2
* Start of sessions page and status

v0.1.1
* udp for eXosip2 messaging support
* automatic authentication integrated
* Start of core messaging code
* Replacing json with xdp
* Start of contacts (phonebook) page

v0.1.0
* Core authentication complete 
* Core registration with expiration
* Extended registration support for sipwitchqt clients

v0.0.2
* More complete test harness debug config
* Reset and dialogs for client
* Release packaging for primary platforms

v0.0.1
* Initial development testing release



utils/swcert-create.rb
#!/usr/bin/env ruby
# Copyright (C) 2017-2018 Tycho Softworks
#
# This file is free software; as a special exception the author gives
# unlimited permission to copy and/or distribute it, with or without
# modifications, as long as this notice is preserved.

['optparse', 'openssl', 'fileutils'].each {|mod| require mod}

class SelfSignedCertificate
  def initialize(server = "localhost", keysize = 1024, org="SipWitchQt", country = "US")
    @key = OpenSSL::PKey::RSA.new(keysize)
    public_key = @key.public_key

    subject = "/C=#{country}/O=#{org}/CN=#{server}"

    @cert = OpenSSL::X509::Certificate.new
    @cert.subject = @cert.issuer = OpenSSL::X509::Name.parse(subject)
    @cert.not_before = Time.now
    @cert.not_after = Time.now + 365 * 24 * 60 * 60
    @cert.public_key = public_key
    @cert.serial = 0x0
    @cert.version = 2

    ef = OpenSSL::X509::ExtensionFactory.new
    ef.subject_certificate = @cert
    ef.issuer_certificate = @cert
    @cert.extensions = [
        ef.create_extension("basicConstraints","CA:TRUE", true),
        ef.create_extension("subjectKeyIdentifier", "hash"),
    # ef.create_extension("keyUsage", "cRLSign,keyCertSign", true),
    ]
    @cert.add_extension ef.create_extension("authorityKeyIdentifier",
                                           "keyid:always,issuer:always")

    @cert.sign @key, OpenSSL::Digest::SHA1.new
  end

  def public_pem
    @cert.to_pem
  end

  def private_key
    @key
  end
end

keysize = 1024
server = 'localhost'
country = 'US'
organization = "SipWitchQt"
banner = 'Usage: swcert-create [options]'

datadir = '/var/lib/sipwitchqt'
datadir = '../testdata' if File.directory? '../testdata'

OptionParser.new do |opts|
  opts.banner = banner

  opts.on('-4', '--4096', 'specify 4096 bit keysize') do
    keysize = 4096
  end

  opts.on('-c CODE', '--country CODE', 'specify country code') do |code|
    country = code
  end

  opts.on('-o ORG', '--org ORG', 'specify certificate hostname') do |org|
    organization = org
  end

  opts.on('-s HOST', '--server HOST', 'specify certificate hostname') do |host|
    server = host
  end

  opts.on('-u', '--userdata', 'output to userdata directory') do
    datadir = '../userdata'
  end

  opts.on_tail('-h', '--help', 'Show this message') do
    puts opts
    exit
  end
end.parse!
abort(banner) unless ARGV.size == 0

cert = SelfSignedCertificate.new(server, keysize, organization, country)

File.open("#{datadir}/public.pem", "w") do |f|
  f.puts cert.public_pem
end

File.open("#{datadir}/private.key", "w") do |f|
  f.puts cert.private_key
end

FileUtils.chmod 0600, "#{datadir}/private.key"


